src/bin/cargo/main.rs

```diff
@@ -52,18 +53,32 @@ fn builtin_aliases_execs(cmd: &str) -> Option<&(&str, &str, &str)> {
     BUILTIN_ALIASES.iter().find(|alias| alias.0 == cmd)
 }
 
+/// Resolve the aliased command from the [`Config`] with a given command string.
+///
+/// The search fallback chain is:
+///
+/// 1. Get the aliased command as a string.
+/// 2. If the aliased command is not a string, get it as an array.
+/// 3. If cannot find any, finds one insides [`BUILTIN_ALIASES`].
+/// 4. If still no command is found, return `None`.
 fn aliased_command(config: &Config, command: &str) -> CargoResult<Option<Vec<String>>> {
-    let alias_name = format!("alias.{}", command);
-    let user_alias = match config.get_string(&alias_name) {
-        Ok(Some(record)) => Some(
+    let alias_name = format!("alias.{command}");
+    let user_alias = match config.get::<Value<String>>(&alias_name) {
+        Ok(record) => Some(
             record
                 .val
                 .split_whitespace()
                 .map(|s| s.to_string())
                 .collect(),
         ),
-        Ok(None) => None,
-        Err(_) => config.get::<Option<Vec<String>>>(&alias_name)?,
+        Err(e) => match e.downcast_ref::<ConfigError>().map(|e| e.kind()) {
+            // Just report we cannot find anything.
+            Some(ConfigErrorKind::MissingKey) => None,
+            // Alias might be set as an toml array, fallback and try it.
+            Some(ConfigErrorKind::UnexpectedValue) => config.get::<Option<_>>(&alias_name)?,
+            // A unrecoverable error, e.g. TOML parsing error. Relay it to the user.
+            Some(ConfigErrorKind::Custom) | None => return Err(e),
```

Comment:
> Propagate unrecoverable error to users when resolving aliases 

Could you help me understand why this change does this?  I'm not too familiar with the config code and my quick look at it didn't clarify things for me.  From my naive perspective, they look the same
- `MissingKey` looks to be handled by doing `get::<Option<Value<String>>` (ie `get_string`)
- `UnexpectedValue` was handled by falling back to the `Option<Vec<String>>` case on error
- `Custom` was being handled by us falling back to `Option<Vec<String>>` and then that failing

Reply:
Your understanding is correct! The main idea is making parsing error visible to users.

Reply:
>  The main idea is making parsing error visible to users.

Thats what I'm missing is how the behavior differs.  From my looking through, they seem identical.

Reply:
It was these pieces of code that swallow the parse error (`has_key` -> `get_cv`) when deserlizing `Option<…>`

https://github.com/rust-lang/cargo/blob/082503982ea0fb7a8fd72210427d43a2e2128a63/src/cargo/util/config/de.rs#L83-L88
https://github.com/rust-lang/cargo/blob/082503982ea0fb7a8fd72210427d43a2e2128a63/src/cargo/util/config/mod.rs#L696-L699

Reply:
Maybe its just me but it feels like that might be more of the crux of the problem that we are swallowing errors there.

What if we made `has_key` return an error or added a `try_has_key`?